File: android/src/main/kotlin/com/live_location/LiveLocationPlugin.kt
package com.live_location

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.annotation.NonNull
import androidx.core.app.ActivityCompat
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry

/// Main plugin class for Live Location tracking on Android.
///
/// Manages the lifecycle of location tracking, permissions, and communication
/// with the Dart layer via method and event channels.
class LiveLocationPlugin : FlutterPlugin, ActivityAware, PluginRegistry.RequestPermissionsResultListener {
    companion object {
        private const val METHOD_CHANNEL = "com.flutter_live_location/methods"
        private const val FOREGROUND_EVENT_CHANNEL = "com.flutter_live_location/foreground_locations"
        private const val BACKGROUND_EVENT_CHANNEL = "com.flutter_live_location/background_locations"
        private const val PERMISSION_REQUEST_CODE = 100
    }

    private lateinit var context: Context
    private var activity: android.app.Activity? = null
    private var activityPluginBinding: ActivityPluginBinding? = null
    private lateinit var methodChannel: MethodChannel
    private lateinit var foregroundEventChannel: EventChannel
    private lateinit var backgroundEventChannel: EventChannel

    private var locationManager: LocationManager? = null
    private var foregroundEventSink: EventChannel.EventSink? = null
    private var backgroundEventSink: EventChannel.EventSink? = null

    private var pendingPermissionResult: MethodChannel.Result? = null

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext

        // Set up method channel
        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        methodChannel.setMethodCallHandler { call, result ->
            handleMethodCall(call, result)
        }

        // Set up event channels
        foregroundEventChannel = EventChannel(binding.binaryMessenger, FOREGROUND_EVENT_CHANNEL)
        foregroundEventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                foregroundEventSink = events
                locationManager?.onStreamListened()
            }

            override fun onCancel(arguments: Any?) {
                foregroundEventSink = null
                if (backgroundEventSink == null) {
                    locationManager?.onStreamCancelled()
                }
            }
        })

        backgroundEventChannel = EventChannel(binding.binaryMessenger, BACKGROUND_EVENT_CHANNEL)
        backgroundEventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                backgroundEventSink = events
                locationManager?.onStreamListened()
            }

            override fun onCancel(arguments: Any?) {
                backgroundEventSink = null
                if (foregroundEventSink == null) {
                    locationManager?.onStreamCancelled()
                }
            }
        })
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        foregroundEventChannel.setStreamHandler(null)
        backgroundEventChannel.setStreamHandler(null)
        locationManager?.dispose()
        activity = null
    }

    // ActivityAware

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityPluginBinding = binding
        binding.addRequestPermissionsResultListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activityPluginBinding?.removeRequestPermissionsResultListener(this)
        activityPluginBinding = null
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityPluginBinding = binding
        binding.addRequestPermissionsResultListener(this)
    }

    override fun onDetachedFromActivity() {
        activityPluginBinding?.removeRequestPermissionsResultListener(this)
        activityPluginBinding = null
        activity = null
    }

    // PluginRegistry.RequestPermissionsResultListener

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ): Boolean {
        if (requestCode != PERMISSION_REQUEST_CODE) return false
        val pending = pendingPermissionResult ?: return false
        pendingPermissionResult = null

        val granted = grantResults.isNotEmpty() &&
            grantResults[0] == PackageManager.PERMISSION_GRANTED
        if (granted) {
            pending.success("granted")
        } else {
            val act = activity
            val showRationale = act != null && ActivityCompat.shouldShowRequestPermissionRationale(
                act, Manifest.permission.ACCESS_FINE_LOCATION
            )
            pending.success(if (showRationale) "denied" else "deniedForever")
        }
        return true
    }

    private fun handleMethodCall(call: io.flutter.plugin.common.MethodCall, result: io.flutter.plugin.common.MethodChannel.Result) {
        try {
            when (call.method) {
                "initialize" -> initialize(call, result)
                "startTracking" -> startTracking(call, result)
                "stopTracking" -> stopTracking(result)
                "onStreamListened" -> {
                    locationManager?.onStreamListened()
                    result.success(null)
                }
                "onStreamCancelled" -> {
                    locationManager?.onStreamCancelled()
                    result.success(null)
                }
                "checkPermission" -> checkPermission(result)
                "requestPermission" -> requestPermission(result)
                "checkLocationServiceEnabled" -> checkLocationServiceEnabled(result)
                "dispose" -> dispose(result)
                else -> result.notImplemented()
            }
        } catch (e: Exception) {
            result.error("PLATFORM_ERROR", e.message, e.stackTraceToString())
        }
    }

    private fun initialize(call: io.flutter.plugin.common.MethodCall, result: io.flutter.plugin.common.MethodChannel.Result) {
        try {
            val sysLocationManager = context.getSystemService(Context.LOCATION_SERVICE)
                as android.location.LocationManager
            if (!sysLocationManager.isLocationEnabled) {
                result.error(
                    "LOCATION_SERVICE_DISABLED",
                    "Location services are disabled on this device. " +
                        "Enable location in device settings before initializing.",
                    null
                )
                return
            }

            val timeInterval = (call.argument<Number>("timeIntervalSeconds") ?: 2).toInt()
            val accuracy = call.argument<String>("accuracy") ?: "high"
            val enableBackground = call.argument<Boolean>("enableBackground") ?: false
            val distanceFilter = (call.argument<Number>("distanceFilterMeters") ?: 0.0).toDouble()

            locationManager = LocationManager(
                context = context,
                timeIntervalSeconds = timeInterval,
                accuracy = accuracy,
                enableBackground = enableBackground,
                distanceFilterMeters = distanceFilter,
                foregroundEventSink = { location ->
                    foregroundEventSink?.success(location)
                },
                backgroundEventSink = { location ->
                    backgroundEventSink?.success(location)
                },
                onError = { code, message ->
                    methodChannel.invokeMethod("onError", mapOf("code" to code, "message" to message))
                }
            )

            result.success(null)
        } catch (e: Exception) {
            result.error("INIT_ERROR", e.message, null)
        }
    }

    private fun startTracking(call: io.flutter.plugin.common.MethodCall, result: io.flutter.plugin.common.MethodChannel.Result) {
        try {
            locationManager?.startTracking()
            result.success(null)
        } catch (e: SecurityException) {
            result.error("PERMISSION_DENIED", e.message, null)
        } catch (e: Exception) {
            result.error("START_TRACKING_ERROR", e.message, null)
        }
    }

    private fun stopTracking(result: io.flutter.plugin.common.MethodChannel.Result) {
        try {
            locationManager?.stopTracking()
            result.success(null)
        } catch (e: Exception) {
            result.error("STOP_TRACKING_ERROR", e.message, null)
        }
    }

    private fun checkPermission(result: MethodChannel.Result) {
        val fineGranted = ActivityCompat.checkSelfPermission(
            context, Manifest.permission.ACCESS_FINE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED
        val coarseGranted = ActivityCompat.checkSelfPermission(
            context, Manifest.permission.ACCESS_COARSE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED
        result.success(if (fineGranted || coarseGranted) "granted" else "denied")
    }

    private fun requestPermission(result: MethodChannel.Result) {
        val fineGranted = ActivityCompat.checkSelfPermission(
            context, Manifest.permission.ACCESS_FINE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED
        if (fineGranted) {
            result.success("granted")
            return
        }

        val act = activity
        if (act == null) {
            result.error("NO_ACTIVITY", "Activity is not available for permission request", null)
            return
        }

        pendingPermissionResult = result
        ActivityCompat.requestPermissions(
            act,
            arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
            PERMISSION_REQUEST_CODE
        )
    }

    private fun checkLocationServiceEnabled(result: MethodChannel.Result) {
        val sysLocationManager = context.getSystemService(Context.LOCATION_SERVICE)
            as android.location.LocationManager
        result.success(sysLocationManager.isLocationEnabled)
    }

    private fun dispose(result: io.flutter.plugin.common.MethodChannel.Result) {
        try {
            locationManager?.dispose()
            locationManager = null
            result.success(null)
        } catch (e: Exception) {
            result.error("DISPOSE_ERROR", e.message, null)
        }
    }
}

File: ios/Classes/LiveLocationPlugin.swift
import Flutter
import CoreLocation

/// Manages foreground location stream
private class ForegroundLocationStreamHandler: NSObject, FlutterStreamHandler {
    var eventSink: FlutterEventSink?
    weak var locationManager: LocationManager?

    func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        self.eventSink = events
        locationManager?.onStreamListened()
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        let plugin = locationManager?.plugin
        if plugin?.backgroundStreamHandler?.eventSink == nil {
            locationManager?.onStreamCancelled()
        }
        return nil
    }
}

/// Manages background location stream
private class BackgroundLocationStreamHandler: NSObject, FlutterStreamHandler {
    var eventSink: FlutterEventSink?
    weak var locationManager: LocationManager?

    func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        self.eventSink = events
        locationManager?.onStreamListened()
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        let plugin = locationManager?.plugin
        if plugin?.foregroundStreamHandler?.eventSink == nil {
            locationManager?.onStreamCancelled()
        }
        return nil
    }
}

/// Main plugin class for Live Location tracking on iOS.
///
/// Manages the lifecycle of location tracking, permissions, and communication
/// with the Dart layer via method and event channels.
public class LiveLocationPlugin: NSObject, FlutterPlugin {
    static let METHOD_CHANNEL = "com.flutter_live_location/methods"
    static let FOREGROUND_EVENT_CHANNEL = "com.flutter_live_location/foreground_locations"
    static let BACKGROUND_EVENT_CHANNEL = "com.flutter_live_location/background_locations"

    private var methodChannel: FlutterMethodChannel?
    private var foregroundEventChannel: FlutterEventChannel?
    private var backgroundEventChannel: FlutterEventChannel?

    private var locationManager: LocationManager?
    fileprivate var foregroundStreamHandler: ForegroundLocationStreamHandler?
    fileprivate var backgroundStreamHandler: BackgroundLocationStreamHandler?


    public static func dummyMethodToEnforceBundling() {
        // This enforces bundling of the class
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = LiveLocationPlugin()
        instance.setupChannels(with: registrar.messenger())
    }

    public func dummyMethodToEnforceBundling() {
        // Do nothing
    }

    // MARK: - FlutterPlugin

    public static func dummy(withIdentifier identifier: String) {}

    func setupChannels(with binaryMessenger: FlutterBinaryMessenger) {
        // Method channel
        methodChannel = FlutterMethodChannel(
            name: Self.METHOD_CHANNEL,
            binaryMessenger: binaryMessenger
        )
        methodChannel?.setMethodCallHandler { [weak self] call, result in
            self?.handle(call, result: result)
        }

        // Foreground event channel with dedicated handler
        foregroundStreamHandler = ForegroundLocationStreamHandler()
        foregroundEventChannel = FlutterEventChannel(
            name: Self.FOREGROUND_EVENT_CHANNEL,
            binaryMessenger: binaryMessenger
        )
        foregroundEventChannel?.setStreamHandler(foregroundStreamHandler)

        // Background event channel with dedicated handler
        backgroundStreamHandler = BackgroundLocationStreamHandler()
        backgroundEventChannel = FlutterEventChannel(
            name: Self.BACKGROUND_EVENT_CHANNEL,
            binaryMessenger: binaryMessenger
        )
        backgroundEventChannel?.setStreamHandler(backgroundStreamHandler)
    }

    // MARK: - Method Call Handler

    public func handle(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        switch call.method {
        case "initialize":
            initialize(call, result: result)
        case "startTracking":
            startTracking(call, result: result)
        case "stopTracking":
            stopTracking(result: result)

        case "onStreamListened":
            locationManager?.onStreamListened()
            result(nil)
        case "onStreamCancelled":
            locationManager?.onStreamCancelled()
            result(nil)
        case "dispose":
            dispose(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func initialize(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        guard CLLocationManager.locationServicesEnabled() else {
            result(FlutterError(
                code: "LOCATION_SERVICE_DISABLED",
                message: "Location services are disabled on this device. " +
                    "Enable location in Settings > Privacy > Location Services.",
                details: nil
            ))
            return
        }

        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(
                code: "INVALID_ARGS",
                message: "Invalid arguments",
                details: nil
            ))
            return
        }

        do {
            let timeInterval = args["timeIntervalSeconds"] as? Int ?? 2
            let accuracy = args["accuracy"] as? String ?? "high"
            let enableBackground = args["enableBackground"] as? Bool ?? false
            let distanceFilter = args["distanceFilterMeters"] as? Double ?? 0.0
            locationManager = LocationManager(
                plugin: self,
                timeIntervalSeconds: Int32(timeInterval),
                accuracy: accuracy,
                enableBackground: enableBackground,
                distanceFilterMeters: distanceFilter,
                foregroundEventSink: { [weak self] location in
                    self?.foregroundStreamHandler?.eventSink?(location)
                },
                backgroundEventSink: { [weak self] location in
                    self?.backgroundStreamHandler?.eventSink?(location)
                },
                onError: { [weak self] code, message in
                    self?.methodChannel?.invokeMethod(
                        "onError",
                        arguments: ["code": code, "message": message]
                    )
                }
            )

            // Link stream handlers to location manager
            foregroundStreamHandler?.locationManager = locationManager
            backgroundStreamHandler?.locationManager = locationManager

            result(nil)
        } catch {
            result(FlutterError(
                code: "INIT_ERROR",
                message: error.localizedDescription,
                details: nil
            ))
        }
    }

    private func startTracking(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        do {
            try locationManager?.startTracking()
            result(nil)
        } catch LocationError.permissionDenied {
            result(FlutterError(
                code: "PERMISSION_DENIED",
                message: "Location permission not granted",
                details: nil
            ))
        } catch {
            result(FlutterError(
                code: "START_TRACKING_ERROR",
                message: error.localizedDescription,
                details: nil
            ))
        }
    }

    private func stopTracking(result: @escaping FlutterResult) {
        do {
            locationManager?.stopTracking()
            result(nil)
        } catch {
            result(FlutterError(
                code: "STOP_TRACKING_ERROR",
                message: error.localizedDescription,
                details: nil
            ))
        }
    }



    private func dispose(result: @escaping FlutterResult) {
        do {
            locationManager?.dispose()
            locationManager = nil
            result(nil)
        } catch {
            result(FlutterError(
                code: "DISPOSE_ERROR",
                message: error.localizedDescription,
                details: nil
            ))
        }
    }

    // Helper to forward errors to Dart via the method channel
    private func onError(_ code: String, _ message: String) {
        methodChannel?.invokeMethod(
            "onError",
            arguments: ["code": code, "message": message]
        )
    }


}
